File: src/prev_5week_binary_search_ts/124. Binary Tree Maximum Path Sum.py
class Solution:
  def maxPathSum(self, root: TreeNode) -> int:
    max_path = float("-inf")  # placeholder to be updated

    def get_max_gain(node):
      nonlocal max_path  # This tells that max_path is not a local variable
      if node is None:
        return 0
      # Read the part important observations
      gain_on_left = max(get_max_gain(node.left), 0)
    # Read the part important observations
      gain_on_right = max(get_max_gain(node.right), 0)
      # Read first three images of going down the recursion stack
      current_max_path = node.val + gain_on_left + gain_on_right
      # Read first three images of going down the recursion stack
      max_path = max(max_path, current_max_path)
      # Read the last image of going down the recursion stack
      return node.val + max(gain_on_left, gain_on_right)
    
    get_max_gain(root) # Starts the recursion chain
    return max_path		

File: src/2023/prev_5week_binary_search_ts/124. Binary Tree Maximum Path Sum.ts
class TreeNode {
	val: number;
	left: TreeNode | null;
	right: TreeNode | null;
	constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
		this.val = val === undefined ? 0 : val;
		this.left = left === undefined ? null : left;
		this.right = right === undefined ? null : right;
	}
}

function maxPathSum(root: TreeNode | null): number {
  let result: number = -Infinity;

  function getMaxGain(node: TreeNode) {
    if (node === null) return 0;

    const gainOnLeft = Math.max(getMaxGain(node.left), 0);
    const gainOnRight = Math.max(getMaxGain(node.right), 0);
    const currMaxPath = node.val + gainOnLeft + gainOnRight;
    result = Math.max(result, currMaxPath);

    return node.val + Math.max(gainOnLeft, gainOnRight);
  }
  getMaxGain(root);

  return result;
};